lib/abci/handlers/endBlockHandlerFactory.js

```diff
@@ -68,7 +84,35 @@ function endBlockHandlerFactory(
     const header = blockExecutionContext.getHeader();
     const coreChainLock = latestCoreChainLock.getChainLock();
 
-    let response = {};
+    let newValidatorSet = null;
+    let newValidatorsUpdates = null;
+    // validator set is rotated every ROTATION_BLOCK_INTERVAL blocks
+    if (height % ROTATION_BLOCK_INTERVAL === 0) {
+      const smlStore = simplifiedMasternodeList.getStore();
+      const sml = smlStore.getSMLbyHeight(coreChainLock.height);
+      const validatorLlmqType = sml.getValidatorLLMQType();
+
+      const validatorQuorumHash = await new ValidatorSet(sml).getHash(
+        Buffer.from(header.lastCommitHash),
+      );
+
+      container.register({
+        validatorQuorumHash: asValue(validatorQuorumHash),
+      });
+
+      newValidatorSet = sml.getQuorum(validatorLlmqType, validatorQuorumHash);
+
+      const validators = await getValidatorSetInfo(validatorLlmqType, newValidatorSet.quorumHash);
+      if (!validators || !validators.length) {
+        // this should never happen
+        throw new NotFoundAbciError('No validator info found');
+      }
+
+      newValidatorsUpdates = ValidatorSet.fillValidatorUpdates(validators);
+      consensusLogger.trace(`Starting validator set rotation: new quorumHash ${newValidatorSet.quorumHash}`);
+    }
```

Comment:
Move this logic to `ValidatorSet#rotate(height)`. If validators are rotated you need to return true.
Instead of `ValidatorSet.fillValidatorUpdates(validators);`. You can do `ValidatorSet#toABCIValidatorUpdate()` without passing validators since ValidatorSet already will contain them. Probably you need `Validator` class to represent a specific validator.

Reply:
Done. Needed to change class name to ValidatorQuorums as ValidatorSet is already used by a js-abci type